hack: add otel collector, prometheus, and grafana to compose file

jsternberg opened the pull request:
This includes the otel collector and uses the otel collector to direct
traces and metrics to the correct services. This is intended to act as
more of what a pipeline might look like in a production environment for
the traces and metrics.

Traces are forwarded from the otel collector directly to jaeger similar
to how that was configured before.

Metrics are forwarded to the prometheus exporter which prometheus will
use to scrape the metrics. A grafana instance has been included to use
with the prometheus instance to visualize the data. The default username
and password is `moby/moby` and the URL for the prometheus instance is
`http://prometheus:9090`. This state along with any dashboards will be
retained in your local development environment through a volume.

tonistiigi commented on hack/composefiles/grafana.ini:
```diff
@@ -0,0 +1,3 @@
+[security]
```
Can't this config specify the default data-source(maybe even dashboards) so they don't need to be manually defined.

jsternberg replied:
It's another configuration file that does that. I couldn't figure out why it wasn't working so I removed it. I figured out I was putting the file in the wrong location and it wasn't being picked up so now the auto-provisioning is in there.

tonistiigi reviewed (commented):
I could not get this working without more guide or help in `hack/compose` of how this should be used.

The Jaeger data seems to work, although I didn't figure out how to connect tempo to the opentelemetry collector. I did not see any metrics data, via prometheus or otherwise.

tonistiigi reviewed (commented):
Not directly part of this PR but we should not disable unix socket access to buildkit. I'm not sure why the TCP 1234 port is needed atm as buildx/buildctl should both understand connecting via `docker-container://` scheme.

jsternberg commented:
> I could not get this working without more guide or help in hack/compose of how this should be used.
>
> The Jaeger data seems to work, although I didn't figure out how to connect tempo to the opentelemetry collector. I did not see any metrics data, via prometheus or otherwise.

I've moved the documention in `hack/compose` to a README in `hack/composefiles` and referenced the location in the original script. That README should describe how to use it including the metrics.

The metrics from buildkit are automatically sent, but I can't automatically configure buildx with this so you need to do it using some environment variables. Might be worth revisiting client metrics in the future and do something similar to what happens with traces, but I'd like to sort out the tracing API problems in the buildkit client first before doing that.

> Not directly part of this PR but we should not disable unix socket access to buildkit. I'm not sure why the TCP 1234 port is needed atm as buildx/buildctl should both understand connecting via docker-container:// scheme.

Mostly because I didn't know this was a thing. I've updated the various instructions that use TCP 1234 and changed it so it uses the default unix socket.

crazy-max commented on hack/composefiles/influxdb.conf:
This conf does not seem used anywhere

jsternberg replied:
I was experimenting with prometheus vs influxdb for the metric store. I'm sticking with prometheus for now just because it's more universal, but I'm not a big fan of prometheus' data model and it's difficult to use for non-regular data.

crazy-max reviewed (commented):
Looks great!

But get some issues currently:

___

Network issues when starting a build using this builder:

```
$ docker buildx create \
    --bootstrap \
    --name dev \
    --driver remote \
    docker-container://buildkit-dev

$ docker buildx --builder dev build .
...
...
#17 [git 2/2] RUN apk add --no-cache git
#17 0.104 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/main/x86_64/APKINDEX.tar.gz
#17 5.108 WARNING: fetching https://dl-cdn.alpinelinux.org/alpine/v3.18/main: temporary error (try again later)
#17 5.108 fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/APKINDEX.tar.gz
#17 10.11 WARNING: fetching https://dl-cdn.alpinelinux.org/alpine/v3.18/community: temporary error (try again later)
#17 10.11 ERROR: unable to select packages:
#17 10.11   git (no such package):
#17 10.11     required by: world[git]
#17 ERROR: process "/bin/sh -c apk add --no-cache git" did not complete successfully: exit code: 1

#14 [gobuild-base 1/2] RUN apk add --no-cache file bash clang lld musl-dev pkgconfig git make
#14 10.15 ERROR: unable to select packages:
#14 10.15   bash (no such package):
#14 10.15     required by: world[bash]
#14 10.15   clang (no such package):
#14 10.15     required by: world[clang]
#14 10.15   file (no such package):
#14 10.15     required by: world[file]
#14 10.15   git (no such package):
#14 10.15     required by: world[git]
#14 10.15   lld (no such package):
#14 10.15     required by: world[lld]
#14 10.15   make (no such package):
#14 10.15     required by: world[make]
#14 10.15   musl-dev (no such package):
#14 10.15     required by: world[musl-dev]
#14 10.15   pkgconfig (no such package):
#14 10.15     required by: world[pkgconfig]
#14 ERROR: process "/bin/sh -c apk add --no-cache file bash clang lld musl-dev pkgconfig git make" did not complete successfully: exit code: 8
```

I don't repro with docker driver or another container builder (latest `moby/buildkit:master`).

___

I see the datasource in grafana: http://localhost:3000/connections/datasources

![image](https://github.com/moby/buildkit/assets/1951866/cff042f0-6afe-470e-96c8-36f3310a838d)

But when doing a build no metrics?

![image](https://github.com/moby/buildkit/assets/1951866/65d12269-be1b-4fa6-9090-decc0b388a04)

___

I was also wondering if we could use this stack for integration tests? Or at least be able to check this stack is operational by scraping logs similar to https://github.com/crazy-max/docker-msmtpd/blob/37bfacfed9e465e1a04ddcdbe17cb4d2643edd03/.github/workflows/test.yml#L41-L55?

jsternberg commented:
@crazy-max 

> Network issues when starting a build using this builder:

This is weird. I don't get this when doing this locally and I'm not sure what could be causing it. The compose file uses `moby/buildkit:local` which is the default name for `make images`. If you don't use `--build` with the compose invocation, it'll use whatever was already present. Maybe you have something local from another branch you're accidentally using? Try using `hack/compose up -d --build`.

> But when doing a build no metrics?

Fixed this. I copied the configuration from the docs and it sets the cache level to high which seems to set the caching on the metrics explorer to some ridiculous metric that causes it to not requery the schema. I've changed it to none since this is a local prometheus/grafana instance for development and we can always nuke the volume and recreate it on our local machines if it starts to cause problems.

jsternberg commented:
Just going to close this for now since https://github.com/moby/buildkit/pull/4757 was merged and it does 90% of what I want. At the moment, I don't find the prometheus/grafana part of this useful and they're easy enough to add as a separate component so I don't really feel the need to improve this PR.
